tests/suite/cli_paths.rs

```diff
@@ -415,46 +415,53 @@ mod windows {
     }
 
     #[test]
+    #[ignore = "FIXME: Must evaluate whether this test item should be removed."]
```

Comment:
I'm getting increasingly concerned about dropping the non-UTF8 tests. If you think about it, Rust's `str` is guaranteed to be valid UTF8, but we the string stored in the Windows PATH may not be one, and Rustup must be able to work correctly even if some other applications has added a non-UTF8 path to the PATH, so the right abstraction to use in Microsoft's wrapper should be `OsStr`, not `str`.

`windows-registry` also uses the following translation function when it handles `&str` being passed in:

```str
fn pcwstr<T: AsRef<str>>(value: T) -> Vec<u16> {
    value
        .as_ref()
        .encode_utf16()
        .chain(core::iter::once(0))
        .collect()
}
```

This means the range of string values that we can write to the registry is a subset of the string values that can potentially be set in the registry, which may lead to very bad consequences.

Reply:
> TLDR: I think there's a API mistake in the `windows-registry` library that `winreg` didn't have.

Can you open an issue against upstream to check your reasoning? They're usually responsive.

Reply:
> ... since it uses lossy conversion to UTF8, this simply cannot be a reliable source of truth.

I'm not sure I agree? If the Windows registry API enforces valid UTF-16, then using `String::from_utf16_lossy()` is just the easiest `std` API to turn UTF-16 bytes into a `String`. I can't find clear references in https://learn.microsoft.com/en-us/windows/win32/sysinfo/registry-value-types but I wouldn't be surprised if it's the case.

Reply:
> If the Windows registry API enforces valid UTF-16, then using `String::from_utf16_lossy()` is just the easiest `std` API to turn UTF-16 bytes into a `String`.

@djc Exactly. However, the following statement in https://doc.rust-lang.org/std/ffi/struct.OsString.html makes me wonder if this is an appropriate occasion to interpret that `Vec<u16>` as valid UTF16:

> On Windows, strings are often arbitrary sequences of non-zero 16-bit values, interpreted as UTF-16 when it is valid to do so.

Reply:
@djc Oh, wait, the test data in the aforementioned test case is actually not valid UTF16:

```javascript
console.log("\uD800\u0101\u0000".isWellFormed()) // false
```

Reply:
Just catching up with this. I would only note that `Cargo` only really supports unicode paths (because they're stored as `String` in a lot of places). Tools like git or the linker also don't support broken UTF-16. Which is to say, I don't think this is necessarily a high priority. Something else will almost certainly break on invalid UTF-16 and arguably rustup should never be *writing* broken unicode to the registry even if it can read it.

Reply:
@ChrisDenton I see. Personally I still would like to wait for the final response from `windows-registry`'s side to actually make the decision, but thanks a lot for the information! If they decide that this change (to `OsStr`) is not worth it, let's just go with `&str`.

Reply:
The change seems fine to me. I just don't really see is as a big deal either way and not really a blocker. The only way non-unicode paths can happen on Windows is if something has gone badly wrong already (e.g. a broken conversion algorithm or some such). Unlike Unix, Windows paths are documented as being UTF-16 (the lack of validity checking is a kernel implementation detail). On Unix paths can be in any random encoding so it's more important there.